src/Cdo.chpl

```diff
@@ -38,1601 +38,1561 @@ This module is focused on providing interfaces, base classes, and helpers that a
-            robj = new refType();
-            robj.setConnection(this.getConnection());
-            robj.setQueryBuilder(this.getConnection().table(robj.getTable()));
-            robj = qb.getOneAsRecord(robj);    
-        }
-       //qb.Select().Where(foreign_key,this.__cdo_getFieldName(obj,lk));
-    }
+            return obj;
+        }
 
-    proc HasOne(ref obj:?eltType, type refType, foreign_key:string, local_key:string=""): refType{
-       var robj = new refType();
-       obj.setConnection(this.getConnection());
-       obj.setQueryBuilder(this.getConnection().table(obj.getTable()));       
-       robj.setConnection(this.getConnection());
-       robj.setQueryBuilder(this.getConnection().table(robj.getTable())); 
-       var fk = foreign_key; 
-       var lk = local_key;
-       var qb = robj.getQueryBuilder();
-        
-       if(local_key ==""){
-        lk = obj.getPK();
-       }
-       var qr =  qb.Select().Where(foreign_key,this.__cdo_getFieldName(obj,lk)).Limit(1).Query();
-        robj = qr.getOneAsRecord(robj);
-        return robj;
-    }
+        iter All(type eltType){
 
-    iter BelongsToMany(ref obj:?eltType, type refType, intermediate_table:string, local_key:string, foreign_key:string, remote_key :string= "id"){
-       var robj = new refType();
-       obj.setConnection(this.getConnection());
-       obj.setQueryBuilder(this.getConnection().table(obj.getTable()));       
-       robj.setConnection(this.getConnection());
-       robj.setQueryBuilder(this.getConnection().table(robj.getTable()));
-       var qb = robj.getQueryBuilder();
-       var rk = remote_key;
-     
-       var qr = qb.Select([robj.getTable()+".*"])
-      .Join(intermediate_table,robj.getTable()+"."+robj.getPK(),intermediate_table+"."+foreign_key)
-      .Join(obj.getTable(),obj.getTable()+"."+obj.getPK(),intermediate_table+"."+local_key)
-      .Where(obj.getTable()+"."+obj.getPK(),this.__cdo_getFieldName(obj,obj.getPK()))
-      .Query();
-
-      // writeln(qr.toSql());
-
-       robj = qr.getOneAsRecord(robj);
-        while(robj!=nil){
-            yield robj;
-            robj = new refType();
-            robj.setConnection(this.getConnection());
-            robj.setQueryBuilder(this.getConnection().table(robj.getTable()));
-            robj = qb.getOneAsRecord(robj);    
-        }
-    }
+        var obj: eltType;
+        obj.setConnection(this.getConnection());
+        obj.setQueryBuilder(this.getConnection().table(obj.getTable()));
+        
+        var qb = obj.getQueryBuilder().Select()
+            .From(obj.getTable()).Query();
 
-    proc __cdo_getFieldName(ref obj:?eltType, fieldname:string):string{
-        for param i in 1..numFields(eltType) {
-            var fname = getFieldName(eltType,i);
-            if(fieldname == fname){
-                return getFieldRef(obj, i):string; 
+            obj = qb.getOneAsRecord(obj);
+            while(obj!=nil){
+                yield obj;
+                obj = nil;
```

Comment:
I'm having a hard time following this bit. We're setting `obj` to `nil` but then calling a method on it in the next line. Wouldn't that generate a dereference nil error at runtime?

Reply:
The Cdo module is not getting used as much and hasn't been updated in many places. These changes are due to indentation changes